File: javascript/smartlist/smartlist-api-mysql-project/src/modules/assessment-block/assessment-block.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { AssessmentBlockRepository } from './assessment-block.repository';
import { ReorderReqDto } from './dto/request/reorder.req.dto';
import { AssessmentBlockReqDto } from './dto/request/assessment-block.req.dto';
import { AssessmentBlockRespDto } from './dto/response/assessment-block.resp.dto';
import { AssessmentRepository } from '../assessment/assessment.repository';
import { Assessment } from 'src/entities/assessment.entity';
import { AssessmentBlockResponseCodes } from './assessment-block.response.codes';
import { AssessmentBlock } from '../../entities/assessment-block.entity';
import { AssessmentStatus } from 'src/enums/assessment.status';

@Injectable()
export class AssessmentBlockService {
  constructor(
    @InjectRepository(AssessmentBlockRepository)
    private assessmentBlockRepository: AssessmentBlockRepository,
    @InjectRepository(AssessmentRepository)
    private assessmentRepository: AssessmentRepository,
  ) {}

  async getAllByAssessmentId(
    assessmentId: number,
  ): Promise<AssessmentBlockRespDto[]> {
    const assessment: Assessment = await this.assessmentRepository.findOne(
      assessmentId,
    );
    this.throwIfAssessmentNotDefined(assessment);

    const assessmentBlocks: AssessmentBlock[] =
      await this.assessmentBlockRepository.findAllByAssessmentWithQuestions(
        assessment,
      );

    const assessmentBlockRespDtos: AssessmentBlockRespDto[] = [];

    assessmentBlocks.forEach((assessmentBlock: AssessmentBlock) => {
      assessmentBlockRespDtos.push(new AssessmentBlockRespDto(assessmentBlock));
    });
    return assessmentBlockRespDtos;
  }

  async getAssessmentBlockById(
    assessmentBlockId: number,
  ): Promise<AssessmentBlockRespDto> {
    const assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findByIdWithQuestions(
        assessmentBlockId,
      );

    this.throwIfAssessmentBlockNotDefined(assessmentBlock);

    const assessmentBlockRespDto: AssessmentBlockRespDto =
      new AssessmentBlockRespDto(assessmentBlock);
    return assessmentBlockRespDto;
  }

  async createAssessmentBlock(
    assessmentId: number,
    assessmentBlockReqDto: AssessmentBlockReqDto,
  ): Promise<number> {
    const assessment: Assessment = await this.assessmentRepository.findOne(
      assessmentId,
    );
    this.throwIfAssessmentNotDefined(assessment);
    this.throwIfAssessmentNotEditable(assessment);

    let sortOrder: number =
      await this.assessmentBlockRepository.getMaxSortOrder(assessmentId);
    sortOrder = sortOrder != null ? sortOrder + 1 : 1;

    let assessmentBlock: AssessmentBlock =
      this.getEntityFromAssessmentBlockReqDto(assessmentBlockReqDto);
    assessmentBlock.assessment = assessment;
    assessmentBlock.sortOrder = sortOrder;

    assessmentBlock = await this.assessmentBlockRepository.save(
      assessmentBlock,
    );
    return assessmentBlock.id;
  }

  async updateAssessmentBlock(
    assessmentBlockId: number,
    assessmentBlockReqDto: AssessmentBlockReqDto,
  ) {
    let assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findByIdWithAssessment(
        assessmentBlockId,
      );

    this.throwIfAssessmentBlockNotDefined(assessmentBlock);
    this.throwIfAssessmentNotEditable(assessmentBlock.assessment);

    assessmentBlock = this.getEntityFromAssessmentBlockReqDto(
      assessmentBlockReqDto,
    );
    assessmentBlock.id = assessmentBlockId;
    await this.assessmentBlockRepository.save(assessmentBlock);
  }

  async reorderAssessmentBlock(
    assessmentId: number,
    reorderReqDto: ReorderReqDto,
  ) {
    const assessment: Assessment = await this.assessmentRepository.findOne(
      assessmentId,
    );
    this.throwIfAssessmentNotDefined(assessment);
    this.throwIfAssessmentNotEditable(assessment);

    if (reorderReqDto.ids.length == 0) {
      throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_LIST;
    }

    const assessmentBlocks: AssessmentBlock[] =
      await this.assessmentBlockRepository.findAllByAssessment(assessment);
    if (assessmentBlocks.length != reorderReqDto.ids.length) {
      throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_LIST;
    }
    assessmentBlocks.forEach((ab) => {
      if (reorderReqDto.ids.indexOf(ab.id) == -1) {
        throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_LIST;
      }
    });

    reorderReqDto.ids.forEach((id, index) => {
      const assessmentBlock: AssessmentBlock = assessmentBlocks.find(
        (ab) => ab.id == id,
      );
      if (assessmentBlock != null) {
        assessmentBlock.sortOrder = index + 1;
      }
    });

    await this.assessmentBlockRepository.save(assessmentBlocks);
  }

  async deleteAssessmentBlock(assessmentBlockId: number) {
    const assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findByIdWithAssessment(
        assessmentBlockId,
      );

    this.throwIfAssessmentBlockNotDefined(assessmentBlock);
    this.throwIfAssessmentNotEditable(assessmentBlock.assessment);

    await this.assessmentBlockRepository.remove(assessmentBlock);
    await this.assessmentBlockRepository.updateSortOrder(
      assessmentBlock.sortOrder,
      assessmentBlock.assessment.id,
    );
  }

  private throwIfAssessmentNotDefined(assessment: Assessment) {
    if (assessment == null) {
      throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_ID;
    }
  }

  private throwIfAssessmentBlockNotDefined(assessmentBlock: AssessmentBlock) {
    if (assessmentBlock == null) {
      throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID;
    }
  }

  private throwIfAssessmentNotEditable(assessment: Assessment) {
    if (assessment.status != AssessmentStatus.DRAFT) {
      throw AssessmentBlockResponseCodes.ASSESSMENT_ACTION_DENIED;
    }
  }

  private getEntityFromAssessmentBlockReqDto(
    assessmentBlockReqDto: AssessmentBlockReqDto,
  ): AssessmentBlock {
    const assessmentBlock = new AssessmentBlock();
    assessmentBlock.title = assessmentBlockReqDto.title;
    assessmentBlock.instruction = assessmentBlockReqDto.instruction;
    assessmentBlock.duration = assessmentBlockReqDto.duration;
    assessmentBlock.closingComments = assessmentBlockReqDto.closing_comments;
    return assessmentBlock;
  }
}

File: java/smartlist/src/main/java/com/smartlist/module/assessmentblock/AssessmentBlockService.java
package com.smartlist.module.assessmentblock;

import com.smartlist.enums.AssessmentStatus;
import com.smartlist.model.Assessment;
import com.smartlist.model.AssessmentBlock;
import com.smartlist.module.assessment.AssessmentRepository;
import com.smartlist.module.assessmentblock.dto.request.AssessmentBlockReqDTO;
import com.smartlist.module.assessmentblock.dto.request.ReorderReqDTO;
import com.smartlist.module.assessmentblock.dto.response.AssessmentBlockRespDTO;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
public class AssessmentBlockService {

    //Autowired
    final AssessmentRepository assessmentRepository;
    final AssessmentBlockRepository assessmentBlockRepository;

    //Global
    private final ModelMapper modelMapper;

    public AssessmentBlockService(final AssessmentRepository assessmentRepository, final AssessmentBlockRepository assessmentBlockRepository,
                                  final ModelMapper modelMapper){
        this.assessmentRepository = assessmentRepository;
        this.assessmentBlockRepository = assessmentBlockRepository;
        this.modelMapper = modelMapper;

        AssessmentBlockRespDTO.addEntityToDTOMappings(this.modelMapper);
    }

    public List<AssessmentBlockRespDTO> getAllByAssessmentId(final Integer assessmentId) {
        Optional<Assessment> optionalAssessment = this.assessmentRepository.findById(assessmentId);
        Assessment assessment = optionalAssessment.orElseThrow (this::throwIfAssessmentNotDefined);

        List<AssessmentBlock> assessmentBlocks = this.assessmentBlockRepository.findAllByAssessmentWithQuestions(assessment);
        List<AssessmentBlockRespDTO> assessmentBlockRespDTOS = new ArrayList<>();

        assessmentBlocks.forEach(assessmentBlock -> assessmentBlockRespDTOS.add(this.modelMapper.map(assessmentBlock, AssessmentBlockRespDTO.class)));
        return assessmentBlockRespDTOS;
    }


    public AssessmentBlockRespDTO getAssessmentBlockById(final Integer assessmentBlockId) {
        Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findByIdWithQuestions(assessmentBlockId);
        AssessmentBlock assessmentBlock = optionalAssessmentBlock.orElseThrow (this::throwIfAssessmentBlockNotDefined);

        return this.modelMapper.map(assessmentBlock, AssessmentBlockRespDTO.class);
    }

    public Integer createAssessmentBlock(final Integer assessmentId, final AssessmentBlockReqDTO assessmentBlockReqDto) {
        Optional<Assessment> optionalAssessment  = this.assessmentRepository.findById(assessmentId);
        Assessment assessment = optionalAssessment.orElseThrow (this::throwIfAssessmentNotDefined);

        throwIfAssessmentNotEditable(assessment);

        Integer sortOrder = this.assessmentBlockRepository.getMaxSortOrder(assessmentId);
        sortOrder = sortOrder != null ? sortOrder + 1 : 1;

        AssessmentBlock assessmentBlock  = this.modelMapper.map(assessmentBlockReqDto, AssessmentBlock.class);
        assessmentBlock.setAssessment(assessment);
        assessmentBlock.setSortOrder(sortOrder);

        this.assessmentBlockRepository.save(assessmentBlock);
        return assessmentBlock.getId();
    }

    public void updateAssessmentBlock(final Integer assessmentBlockId, final AssessmentBlockReqDTO assessmentBlockReqDto) {
        Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findById(assessmentBlockId);

        AssessmentBlock assessmentBlock = optionalAssessmentBlock.orElseThrow (this::throwIfAssessmentBlockNotDefined);
        throwIfAssessmentNotEditable(assessmentBlock.getAssessment());

        this.modelMapper.map(assessmentBlockReqDto, assessmentBlock);
        this.assessmentBlockRepository.save(assessmentBlock);
    }

    public void reorderAssessmentBlock(final Integer assessmentId, final ReorderReqDTO reorderReqDto) {
        Optional<Assessment> optionalAssessment =  this.assessmentRepository.findById(assessmentId);
        Assessment assessment = optionalAssessment.orElseThrow (this::throwIfAssessmentNotDefined);

        throwIfAssessmentNotEditable(assessment);

        if (reorderReqDto.getIds().isEmpty()) {
            throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_LIST;
        }

        List<AssessmentBlock> assessmentBlocks = this.assessmentBlockRepository.findAllByAssessment(assessment);

        if (assessmentBlocks.size() != reorderReqDto.getIds().size()) {
            throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_LIST;
        }

        assessmentBlocks.forEach(assessmentBlock -> {
            if (!reorderReqDto.getIds().contains(assessmentBlock.getId())) {
                throw AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_LIST;
            }
        });

        for(int i =0; i < reorderReqDto.getIds().size();i++){
            int id = reorderReqDto.getIds().get(i);
            Optional<AssessmentBlock> optionalAssessmentBlock = assessmentBlocks.stream().filter(ab -> ab.getId() == id).findFirst();
            if(optionalAssessmentBlock.isPresent()){
                optionalAssessmentBlock.get().setSortOrder(i + 1);
            }
        }
        this.assessmentBlockRepository.saveAll(assessmentBlocks);

    }

    @Transactional
    public void deleteAssessmentBlock(final Integer assessmentBlockId) {
        Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findById(assessmentBlockId);
        AssessmentBlock assessmentBlock = optionalAssessmentBlock.orElseThrow (this::throwIfAssessmentBlockNotDefined);

        this.throwIfAssessmentNotEditable(assessmentBlock.getAssessment());

        this.assessmentBlockRepository.delete(assessmentBlock);
        this.assessmentBlockRepository.updateSortOrder(assessmentBlock.getSortOrder(), assessmentBlock.getAssessment().getId());
    }


    private void throwIfAssessmentNotEditable(Assessment assessment) {
        if (assessment.getStatus() != AssessmentStatus.DRAFT) {
            throw AssessmentBlockResponseCodes.ASSESSMENT_ACTION_DENIED;
        }
    }

    private ResponseStatusException throwIfAssessmentBlockNotDefined() {
        return AssessmentBlockResponseCodes.INVALID_ASSESSMENT_BLOCK_ID;
    }

    private ResponseStatusException throwIfAssessmentNotDefined() {
        return AssessmentBlockResponseCodes.INVALID_ASSESSMENT_ID;
    }
}
